### PR TITLE
fix normalization in categorical.py

### PR DESCRIPTION
**Pull request by MrNeither**
I suggest this option to fix the problem #43138

### CODE REVIEW

**Review by neerajprad — COMMENTED** [review]
While we are effectively casting `probs` to float type, we still assume that `logits` are provided as floats since `logsumexp` might fail otherwise. Rather than being more permissive and having to do this casting across different distributions, my suggestion would be to instead raise errors if `probs` or `logits` is an integer/long tensor, which is a very reasonable restriction to have and also placed by other torch functions like `torch.multinomial`.

**Comment by MrNeither** [comment]
@neerajprad  I agree that you need to do an integer check for logits. But in my case, I initialize the distribution class using probs, and in case of prohibiting the call using LongTensor, I will have to use double conversions to float. This does not make sense for a categorical distribution, since normalization of probs results in a float, so there can be no further errors. If you prohibit using LongTensor, I give an example of Genetic algorithms where speed is of great importance, this method is called a huge number of times in the selection method. A longtensor ban forces the tensor to float before calling the class.

The difference between the categorical distribution and the rest is that it is sampling from roulette, and here it makes sense to use LongTensor.

I understand that your framework is mainly used for machine learning, but it is also very convenient for discrete optimization tasks (such as GA), and I am very sad to see that you are gradually restricting functions.

If you can't leave the use of LongTensor in probs, I ask you to at least add the _need_normilize_ parameter (or something like that), which will not do normalize, thus giving you a chance to avoid double casting to float.

Thank you very much for your work

**Comment by neerajprad** [comment]
> The difference between the categorical distribution and the rest is that it is sampling from roulette, and here it makes sense to use LongTensor.

The output is a `LongTensor` but I do not see the argument as to why we should treat the input `probs` argument as any different here vs in other distribution classes or PyTorch functions like `torch.multinomial`. I only wanted to point out that this is making an exception for a very specific use case. Other distribution classes (and torch functions) will assume that `probs` is a float tensor. That said, this assumption is implicit at this point, and we can either cast/process to float (as you are doing here) or enforce types more strictly like `torch.multinomial` does. @fritzo, @ngimel - do you have any thoughts on this?

**Comment by ngimel** [comment]
If the concern with requiring float `probs` is primary performance, I think it's misguided. What happens under the hood (on cpu) when you do 
```
self.probs = probs.true_divide(probs.sum(-1, keepdim=True))
```
is 
1) probs are converted to float explicitly
2) probs.sum() is converted to float explicitly
3) division is performed in float and float result is returned. 
gpu does those conversions implicitly, but they still take cycles, and probs.sum() would still be computed with Long inputs. Keep in mind that typically operations on Long are much slower, so performance-wise you are better off converting to float as early as possible. 
If Long->Float conversion is unreasonably slow, that's a separate issue that we should investigate.

**Comment by MrNeither** [comment]
> The output is a `LongTensor` but I do not see the argument as to why we should treat the input `probs` argument as any different here vs in other distribution classes or PyTorch functions like `torch.multinomial`. 

I can give you an example. The roulette method samples from a set according to some probability. This probability is calculated by values. Yes, if it's an quality score or something like that, then float makes sense. But for example, I want to sample values according to their frequency. I can do my own normalization and get a float, but since normalization is built in in the Categorical class, it is best for me (so that there is no double normalization) to set the number of occurrence of objects. And this is an integer value.

**Comment by fritzo** [comment]
@MrNeither just to be clear, we plan to continue to allow integer-valued inputs to `Categorical`, but your code will be faster if you store those integer values in a `FloatTensor`, and your code will be no slower if you store those integer values in a `LongTensor` and then cast via `.float()` before passing to `Categorical(probs=__)`.

**Comment by MrNeither** [comment]
@fritzo Thank you very much for your help! I figured out this situation, it's my mistake. Speed measurements are posted here #43138
